### PR TITLE
World exporter

### PR DESCRIPTION
**Pull request by gonzodepedro**
Adds a plugin to export worlds to meshes

Requires: ignitionrobotics/ign-common#133

Signed-off-by: Gonzalo de Pedro <gonzalo@depedro.com.ar>

### CODE REVIEW

File: src/systems/world_exporter/WorldExporter.hh [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace systems
+{
+  // Forward declarations.
+  class WorldExporterPrivate;
+
+  /// \brief A plugin that exports a model to mesh.
```
Comment (chapulina) [inline src/systems/world_exporter/WorldExporter.hh:36]:
```suggestion
  /// \brief A plugin that exports a world to a mesh.
```

File: src/systems/world_exporter/WorldExporter.cc [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+#include "WorldExporter.hh"
+
+using namespace ignition;
+using namespace gazebo;
+using namespace systems;
+
+
+class ignition::gazebo::systems::WorldExporterPrivate
+{
+  private:
```
Comment (chapulina) [inline src/systems/world_exporter/WorldExporter.cc:55]: Please document all members and classes.

Also, per our style, have one `private` per variable.

File: src/systems/world_exporter/WorldExporter.cc [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    bool rendered{false};
+    common::Mesh worldMesh;
+    Entity world;
+    std::map<Entity, std::pair<math::Pose3d, Entity>> poses;
+    std::vector<math::Matrix4d> subMeshMatrix;
+    common::ColladaExporter exporter;
+
+  private: math::Pose3d GetPose(Entity parent_, math::Pose3d currentPose)
```
Comment (chapulina) [inline src/systems/world_exporter/WorldExporter.cc:63]: How about using `worldPose`?

https://github.com/ignitionrobotics/ign-gazebo/blob/a5abacfa74db6e8db0218bd32c55980e15cf249b/include/ignition/gazebo/Util.hh#L37-L42

File: src/systems/world_exporter/WorldExporter.cc [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+  public: void Export(const EntityComponentManager &_ecm)
+  {
+    if (this->rendered) return;
+
+    _ecm.Each<components::World, components::Name>(
+      [&](const Entity & _entity,
+        const components::World *,
+        const components::Name * _name)->bool
+    {
+      world = _entity;
```
Comment (chapulina) [inline src/systems/world_exporter/WorldExporter.cc:95]:
```suggestion
      this->world = _entity;
```

File: src/systems/world_exporter/WorldExporter.cc [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+    });
+
+    _ecm.Each<components::Model, components::Pose,
+            components::ParentEntity>(
+      [&](const Entity &_entity,
+          const components::Model *,
+          const components::Pose *_pose,
+          const components::ParentEntity *_parent)->bool
+    {
+      poses.insert(std::pair<Entity, std::pair<math::Pose3d, Entity>>(
```
Comment (chapulina) [inline src/systems/world_exporter/WorldExporter.cc:107]:
```suggestion
      this->poses.insert(std::pair<Entity, std::pair<math::Pose3d, Entity>>(
```

File: src/systems/world_exporter/WorldExporter.cc [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+    });
+
+    _ecm.Each<components::Link, components::Pose,
+            components::ParentEntity>(
+      [&](const Entity &_entity,
+          const components::Link *,
+          const components::Pose *_pose,
+          const components::ParentEntity *_parent)->bool
+    {
+      poses.insert(std::pair<Entity, std::pair<math::Pose3d, Entity>>(
```
Comment (chapulina) [inline src/systems/world_exporter/WorldExporter.cc:120]:
```suggestion
      this->poses.insert(std::pair<Entity, std::pair<math::Pose3d, Entity>>(
```

File: src/systems/world_exporter/WorldExporter.cc [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+    {
+      std::string name = _name->Data().empty() ? std::to_string(_entity) :
+          _name->Data();
+
+      igndbg << "====================== Starting: " << name << " - "
+        << _entity << "  ====================== " << std::endl;
+
+      math::Pose3d localPose = this->GetPose(_parent->Data(), _pose->Data());
+
+      common::MaterialPtr mat(new common::Material());
```
Comment (chapulina) [inline src/systems/world_exporter/WorldExporter.cc:149]: This is slightly more efficient:

```suggestion
      common::MaterialPtr mat = std::make_shared<common::Material>();
```

File: src/systems/world_exporter/WorldExporter.cc [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+      auto material = _ecm.Component<components::Material>(_entity);
+      if (material != nullptr)
+      {
+        mat->SetDiffuse(material->Data().Diffuse());
+        mat->SetAmbient(material->Data().Ambient());
+        mat->SetEmissive(material->Data().Emissive());
+        mat->SetSpecular(material->Data().Specular());
+        mat->SetTransparency(_transparency->Data());
```
Comment (chapulina) [inline src/systems/world_exporter/WorldExporter.cc:157]: Could the transparency be set outside of this if?

File: src/systems/world_exporter/WorldExporter.cc [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+
+            subm.lock()->Scale(_geom->Data().MeshShape()->Scale());
+            subMeshMatrix.push_back(matrix);
+
+            igndbg << ">>>>>>>>>" << std::endl;
+          }
+      }
+      else
+      {
+        ignerr << "It is NOT a mesh!"  << std::endl;
```
Comment (chapulina) [inline src/systems/world_exporter/WorldExporter.cc:352]: This message can be a bit misleading. How about saying that `_geom->Data().Type()` can't be exported? I also recommend changing it to a warning.

File: src/systems/world_exporter/WorldExporter.cc [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+
+
+class ignition::gazebo::systems::WorldExporterPrivate
+{
+  private:
+    bool rendered{false};
+    common::Mesh worldMesh;
+    Entity world;
+    std::map<Entity, std::pair<math::Pose3d, Entity>> poses;
+    std::vector<math::Matrix4d> subMeshMatrix;
```
Comment (chapulina) [inline src/systems/world_exporter/WorldExporter.cc:60]: Do these variables need to be member variables? Many of them only seem to be used within the scope of the `Export` function, I think they could just be defined there.

File: src/systems/world_exporter/WorldExporter.cc [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+
+using namespace ignition;
+using namespace gazebo;
+using namespace systems;
+
+
+class ignition::gazebo::systems::WorldExporterPrivate
+{
+  private:
+    bool rendered{false};
```
Comment (chapulina) [inline src/systems/world_exporter/WorldExporter.cc:56]: How about `exported` instead of `rendered`? It seems to match better what is being done.

File: src/systems/world_exporter/WorldExporter.cc [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+      return true;
+    });
+
+    exporter.Export(&this->worldMesh, "./" + this->worldMesh.Name(), true,
+                    subMeshMatrix);
+    this->rendered = true;
+  }
+};
+
+WorldExporter::WorldExporter() : System(),
```
Comment (chapulina) [inline src/systems/world_exporter/WorldExporter.cc:364]: Style nitpick, all functions should have 50 `/` before them:

```suggestion
//////////////////////////////////////////////////
WorldExporter::WorldExporter() : System(),
```

**Review by chapulina — COMMENTED** [review]
I haven't tried it yet. Just left some high-level comments. I also believe we need an `ign-common3` release.

**Comment by nkoenig** [comment]
Would it be possible to add a short tutorial that describes what the world exporter is and how to use it?

**Comment by gonzodepedro** [comment]
> Would it be possible to add a short tutorial that describes what the world exporter is and how to use it?

That is a good idea. Should I add it as a Readme file in the system plugin directory or is there another better place to add it?

**Comment by chapulina** [comment]
> Should I add it as a Readme file in the system plugin directory or is there a better place to add it?

Could you create a new one in the tutorials folder?

https://github.com/ignitionrobotics/ign-gazebo/tree/ign-gazebo4/tutorials

Be sure to link to it from [tutorials.md.in](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/tutorials.md.in). Those tutorials are generated when we do `make doc` and are rendered [here](https://ignitionrobotics.org/api/gazebo/4.0/tutorials.html).

Thanks!

**Comment by nkoenig** [comment]
@gonzodepedro , I added a sample world that you could use in the tutorial. See: 63c9eb7

File: examples/worlds/world_exporter.sdf [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+
+This world will load the WorldExporter plugin, which is designed to convert
+a world to a single Collada file. When this world is run, a subdirectory
+named after the world is created. Within this subdirectory is the mesh and
+materials for the world. After a world with the SDF WorldExporter is run,
+just press CTRL-C to quit Ignition Gazebo.
+
+This example just exports simple shapes into a Collada file. To run use:
+
+    ign gazebo -v 4 world_exporter.sdf
```
Comment (chapulina) [inline examples/worlds/world_exporter.sdf:12]: I suggest running only the server, because there's no `SceneBroadcaster`:

```suggestion
    ign gazebo -v 4 -s world_exporter.sdf
```

**Review by chapulina — COMMENTED** [review]
Nice, it works for me!

@gonzodepedro , besides the tutorial, would it be possible to add a test under `test/integration`? At a minimum, I think it could load a world that has one of each geometry type, and at the end we check that a collada file was generated (and delete it).

File: src/systems/world_exporter/WorldExporter.cc [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+      common::MaterialPtr mat = std::make_shared<common::Material>();
+      auto material = _ecm.Component<components::Material>(_entity);
+      if (material != nullptr)
+      {
+        mat->SetDiffuse(material->Data().Diffuse());
+        mat->SetAmbient(material->Data().Ambient());
+        mat->SetEmissive(material->Data().Emissive());
+        mat->SetSpecular(material->Data().Specular());
+      }
+      mat->SetTransparency(_transparency->Data());
```
Comment (azeey) [inline src/systems/world_exporter/WorldExporter.cc:103]: According to [this](https://www.khronos.org/files/collada_spec_1_4.pdf) Collada spec, a transparency of 0 means fully transparent, but in [SDFormat <transparency>](http://sdformat.org/spec?ver=1.7&elem=visual#visual_transparency), a value of 0 means fully opaque. So, I think we need to do `mat->SetTransparency(1 - _transparency->Data());` here.

As a quick test, I ran the example file and imported the generated dae file into Blender and all the shapes were invisible. Making the change I suggested above fixed it. I also tried running one of the cave worlds, but in that case the changes I suggested did not fix the issue, making me think there's probably another place to set transparency when meshes are involved.

File: src/systems/world_exporter/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,6 @@
+gz_add_system(world-exporter
```
Comment (azeey) [inline src/systems/world_exporter/CMakeLists.txt:1]: Unless we have the intention of making this a generic world exporter, could this be renamed to `collada-world-exporter` or something similar? I think it would be good to have the word `collada` in there.

File: src/systems/world_exporter/WorldExporter.cc [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+}
+
+/////////////////////////////////////////////////
+WorldExporter::~WorldExporter() = default;
+
+/////////////////////////////////////////////////
+void WorldExporter::PostUpdate(const UpdateInfo & /*_info*/,
+    const EntityComponentManager &_ecm)
+{
+  this->dataPtr->Export(_ecm);
```
Comment (azeey) [inline src/systems/world_exporter/WorldExporter.cc:323]: The fact that the world is exported immediately after the system is loaded might not work for the SubT use case. This is because the robots are loaded via the `/world/<world name>create_multiple` ign-transport service which would most likely happen after `WorldExporter::PostUpdate` is called the first time. My suggestion would be to create an ign-transport service in this system that triggers the export.

If dynamically loaded models need not be in the exported mesh, please ignore this comment.

**Review by azeey — CHANGES_REQUESTED** [review]
This is awesome! I was able to generate a cave world and import it into blender. There is an issue related to transparency that I mention below that caused everything to be transparent, but otherwise, the world, with 524 submeshes loaded without a problem 🎉 .

**Comment by chapulina** [comment]
`ign-common 3.8.0` is out. Bumped the requirement in 9ff6f09623745064a0249295cb58cf50e8c14ef6.

File: src/systems/world_exporter/WorldExporter.cc (hunk 14 above)
Comment (gonzodepedro) [reply src/systems/world_exporter/WorldExporter.cc:106]: I'm a little confused. How do you see transparency into blender? I'm seeing this:
blender1

Comment (azeey) [reply src/systems/world_exporter/WorldExporter.cc:106]: I actually used blender 2.9 and this is how it shows up
image

In 2.79, switching to the Material mode 
image

shows the colors are not solid. 
image

File: src/systems/world_exporter/WorldExporter.cc [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+          // for the normal vector.
+          mesh = meshManager->MeshByName("unit_plane");
+
+          scale.X() = _geom->Data().PlaneShape()->Size().X();
+          scale.Y() = _geom->Data().PlaneShape()->Size().Y();
+
+          // // The rotation is the angle between the +z(0,0,1) vector and the
+          // // normal, which are both expressed in the local (Visual) frame.
+          math::Vector3d normal = _geom->Data().PlaneShape()->Normal();
+          localPose.Rot().From2Axes(
```
[the remark this answers is not in the document]
Comment (gonzodepedro) [reply src/systems/world_exporter/WorldExporter.cc:177]: My idea is that localPose (now worldPose) already has in it the rotation and translation of the view, so I'll add on top of it the rotation of the normal and that is what gets exported to the matrix.

File: src/systems/world_exporter/WorldExporter.cc (hunk 16 above)
Comment (nkoenig) [reply src/systems/world_exporter/WorldExporter.cc:326]: Dynamically loaded models are not needed for the subt use case.

File: src/systems/world_exporter/WorldExporter.cc (hunk 14 above)
Comment (nkoenig) [reply src/systems/world_exporter/WorldExporter.cc:106]: This is resolved in https://github.com/ignitionrobotics/ign-common/pull/152

File: src/systems/world_exporter/WorldExporter.hh [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace systems
+{
+  // Forward declarations.
+  class WorldExporterPrivate;
+
+  /// \brief A plugin that exports a model to a mesh.
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply src/systems/world_exporter/WorldExporter.hh]: Updated in 29c0a44f5e667191fc0dd80f2ca533de2d4f51c4

File: examples/worlds/world_exporter.sdf (hunk 13 above)
Comment (nkoenig) [reply examples/worlds/world_exporter.sdf]: Updated in 29c0a44f5e667191fc0dd80f2ca533de2d4f51c4

File: src/systems/world_exporter/CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,6 @@
...
+  SOURCES
+    WorldExporter.cc
+  PUBLIC_LINK_LIBS
+    ignition-rendering${IGN_RENDERING_VER}::ignition-rendering${IGN_RENDERING_VER}
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply src/systems/world_exporter/CMakeLists.txt]: Removed in 29c0a44f5e667191fc0dd80f2ca533de2d4f51c4

File: src/systems/world_exporter/WorldExporter.cc [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+      const ignition::common::Mesh * mesh;
+      std::weak_ptr<ignition::common::SubMesh> subm;
+      math::Vector3d scale;
+      math::Matrix4d matrix(localPose);
+      ignition::common::MeshManager *meshManager =
+          ignition::common::MeshManager::Instance();
+
+      if (_geom->Data().Type() == sdf::GeometryType::BOX)
+      {
+        igndbg << "It is a Box!"  << std::endl;
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply src/systems/world_exporter/WorldExporter.cc]: Removed extra debug statements in 29c0a44f5e667191fc0dd80f2ca533de2d4f51c4

File: src/systems/world_exporter/WorldExporter.cc [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+          igndbg << "Box Found..."  << std::endl;
+          mesh = meshManager->MeshByName("unit_box");
+          scale = _geom->Data().BoxShape()->Size();
+          int i = worldMesh.AddMaterial(mat);
+
+          subm = worldMesh.AddSubMesh(
+            *mesh->SubMeshByIndex(0).lock().get());
+          subm.lock()->SetMaterialIndex(i);
+          subm.lock()->Scale(scale);
+          subMeshMatrix.push_back(matrix);
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply src/systems/world_exporter/WorldExporter.cc]: Added a lambda function in 29c0a44f5e667191fc0dd80f2ca533de2d4f51c4

File: src/systems/world_exporter/WorldExporter.cc [hunk 22]
```diff
@@ -0,0 +1,331 @@
...
+        const components::Geometry *_geom,
+        const components::Transparency *_transparency)->bool
+    {
+      std::string name = _name->Data().empty() ? std::to_string(_entity) :
+          _name->Data();
+
+      igndbg << "====================== Starting: " << name << " - "
+        << _entity << "  ====================== " << std::endl;
+
+      math::Pose3d localPose = gazebo::worldPose(_entity, _ecm);
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply src/systems/world_exporter/WorldExporter.cc]: Fixed in 29c0a44f5e667191fc0dd80f2ca533de2d4f51c4

File: src/systems/world_exporter/WorldExporter.cc [hunk 23]
```diff
@@ -0,0 +1,331 @@
...
+
+          if (!mesh) {
+            ignerr << "MESH NOT FOUND!" << std::endl;
+            return true;
+          }
+
+          for (unsigned int k = 0; k < mesh->SubMeshCount(); k++)
+          {
+            igndbg << ">>>>>>>>> SUBMESH : " << k << std::endl;
+            int j = mesh->SubMeshByIndex(k).lock()->MaterialIndex();
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply src/systems/world_exporter/WorldExporter.cc]: Fixed in 29c0a44f5e667191fc0dd80f2ca533de2d4f51c4

File: src/systems/world_exporter/WorldExporter.cc [hunk 24]
```diff
@@ -0,0 +1,331 @@
...
+            igndbg << "Material index on worldmesh: " <<
+              worldMesh.IndexOfMaterial(mesh->MaterialByIndex(
+                mesh->SubMeshByIndex(k).lock()->MaterialIndex()).get())
+                << std::endl;
+
+            int i = 0;
+            if ( j != -1 )
+            {
+              i = worldMesh.IndexOfMaterial(mesh->MaterialByIndex(
+                mesh->SubMeshByIndex(k).lock()->MaterialIndex()).get());
```
[the remark this answers is not in the document]
Comment (nkoenig) [reply src/systems/world_exporter/WorldExporter.cc]: Fixed in 29c0a44f5e667191fc0dd80f2ca533de2d4f51c4

File: src/systems/world_exporter/WorldExporter.cc (hunk 14 above)
Comment (nkoenig) [reply src/systems/world_exporter/WorldExporter.cc:103]: opaque

File: src/systems/world_exporter/WorldExporter.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (azeey) [reply src/systems/world_exporter/WorldExporter.cc]: > I'll add on top of it the rotation of the normal and that is what gets exported to the matrix

Yes, I think that's what we want, but I don't think that's what's happening. The current code is
```c++
worldPose.Rot().From2Axes(
    math::Vector3d::UnitZ, normal.Normalized());

matrix = math::Matrix4d(worldPose);
```
I believe this will replace the rotation in `worldPose` with the rotation calculated from the normal. I don't think it applies it on top of the current rotation in `worldPose`. Here's my suggestion

```c++
math::Quaterniond normalRot;
normalRot.From2Axes(math::Vector3d::UnitZ, normal.Normalized());
worldPose.Rot() = worldPose.Rot() * normalRot;
```

**Comment by azeey** [comment]
This was automatically closed when  ignitionrobotics/ign-common#152  got merged. Reopening.

File: src/systems/world_exporter/WorldExporter.cc (hunk 14 above)
Comment (nkoenig) [reply src/systems/world_exporter/WorldExporter.cc:103]: The fix in ign-common has been released.

File: src/systems/world_exporter/CMakeLists.txt (hunk 15 above)
Comment (nkoenig) [reply src/systems/world_exporter/CMakeLists.txt]: Renamed in 8d8a7f46cb0816ecdb56ad91e789a86c324039e2 and b42e9044273b77a7b41b48e611ef80f2bbc64556

**Comment by nkoenig** [comment]
> > Should I add it as a Readme file in the system plugin directory or is there a better place to add it?
> 
> Could you create a new one in the tutorials folder?
> 
> https://github.com/ignitionrobotics/ign-gazebo/tree/ign-gazebo4/tutorials
> 
> Be sure to link to it from [tutorials.md.in](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/tutorials.md.in). Those tutorials are generated when we do `make doc` and are rendered [here](https://ignitionrobotics.org/api/gazebo/4.0/tutorials.html).
> 
> Thanks!

Tutorial has been added: c1c88b84900e94600476fe6065c5a544fc901ecd

File: src/systems/world_exporter/WorldExporter.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (nkoenig) [reply src/systems/world_exporter/WorldExporter.cc]: Updated in 3206ad83e13221d83349cf5d44655f382f45a8e6

**Comment by nkoenig** [comment]
I believe all that is left is a test.

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc [hunk 25]
```diff
@@ -0,0 +1,259 @@
...
+      else
+      {
+        ignwarn << "Unsupported geometry type" << std::endl;
+      }
+
+      return true;
+    });
+
+    common::ColladaExporter exporter;
+    exporter.Export(&worldMesh, "./" + worldMesh.Name(), true,
```
Comment (azeey) [inline src/systems/collada_world_exporter/ColladaWorldExporter.cc:232]: I know I said to remove debug messages, but a message indicating that the mesh has been successfully exported and its path would be great.

**Review by azeey — COMMENTED** [review]
I left one more minor comment. Otherwise, with the addition of the test, I think this is ready to go.

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc (hunk 25 above)
Comment (nkoenig) [reply src/systems/collada_world_exporter/ColladaWorldExporter.cc:232]: Added a message in 073bc9115c24303835cbec3f1bb7a90a84af5b7b

**Comment by nkoenig** [comment]
Test has been added in db69c561c51297d3e8fd10fd6be321f69651d022. I believe this completes this PR, pending CI.

File: test/worlds/collada_world_exporter.sdf [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+  <world name="collada_world_exporter_box_test">
+
+    <plugin
+      filename="ignition-gazebo-collada-world-exporter-system"
+      name="ignition::gazebo::systems::ColladaWorldExporter">
+    </plugin>
+
+    <model name="box">
+      <pose>0 0 0.5 0 0 0</pose>
+      <link name="box_link">
```
Comment (chapulina) [inline test/worlds/collada_world_exporter.sdf:12]: Can we add a plane to the test?

File: tutorials/collada_world_exporter.md [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+```
+
+2. Run the world using
+```
+ign gazebo -v 4 -s -r --iterations 1 WORLD_FILE_NAME
+```
+
+3. A subdirectory, named after the world, has been created in the current working directory. Within this subdirectory is the mesh and materials for the world.
+
+Refer to the [collada_world_exporter.sdf](https://github.com/ignitionrobotics/ign-gazebo/blob/main/examples/worlds/collada_world_exporter.sdf) example.
```
Comment (chapulina) [inline tutorials/collada_world_exporter.md:25]:
```suggestion
Refer to the [collada_world_exporter.sdf](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/examples/worlds/collada_world_exporter.sdf) example.
```

File: test/integration/collada_world_exporter.cc [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+    EXPECT_FALSE(server->Running());
+  }
+
+  public: std::unique_ptr<Server> server;
+};
+
+/////////////////////////////////////////////////
+TEST_F(ColladaWorldExporterFixture, ExportWorld)
+{
+  this->LoadWorld("test/worlds/collada_world_exporter.sdf");
```
Comment (chapulina) [inline test/integration/collada_world_exporter.cc:56]: We should use `common::joinPaths` here too.

**Review by nkoenig — APPROVED** [review]

File: test/integration/collada_world_exporter.cc [hunk 29]
```diff
@@ -0,0 +1,79 @@
...
+  EXPECT_FALSE(common::exists("./collada_world_exporter_box_test"));
+
+  // Run one iteration which should export the world.
+  server->Run(true, 1, false);
+
+  // The export directory should now exist.
+  EXPECT_TRUE(common::exists("./collada_world_exporter_box_test"));
+
+  // Cleanup
+  common::removeAll("./collada_world_exporter_box_test");
```
Comment (chapulina) [inline test/integration/collada_world_exporter.cc:68]: The test is currently saving the file to the directory where the test is ran from. If there's any problem and the test starts failing, there's a chance these files will be left in the filesystem. I'd recommend at least saving to `PROJECT_BINARY_PATH` like the log test does:

https://github.com/ignitionrobotics/ign-gazebo/blob/8fd9b2647a5c37b1dde5509b5a741c6c955ee10f/test/integration/log_system.cc#L263-L264

In the future, we may use a strategy like the one introduced in https://github.com/ignitionrobotics/ign-common/pull/158.

**Review by chapulina — APPROVED** [review]
LGTM, I just have some minor comments that would be nice to address, but don't need to block the PR.

File: test/integration/collada_world_exporter.cc (hunk 29 above)
Comment (nkoenig) [reply test/integration/collada_world_exporter.cc:68]: I don't think I get set where the test exports the mesh.

File: tutorials/collada_world_exporter.md (hunk 27 above)
Comment (nkoenig) [reply tutorials/collada_world_exporter.md]: Done

File: test/worlds/collada_world_exporter.sdf (hunk 26 above)
Comment (nkoenig) [reply test/worlds/collada_world_exporter.sdf:38]: Added more shapes.

File: test/integration/collada_world_exporter.cc (hunk 29 above)
Comment (nkoenig) [reply test/integration/collada_world_exporter.cc:72]: I added a cleanup call before the exporter starts.

File: test/integration/collada_world_exporter.cc (hunk 28 above)
Comment (nkoenig) [reply test/integration/collada_world_exporter.cc]: Added

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc [hunk 30]
```diff
@@ -0,0 +1,261 @@
...
+            if (i < 0)
+            {
+              i = worldMesh.AddMaterial(mesh->MaterialByIndex(j));
+            }
+          }
+          else
+          {
+            i = worldMesh.AddMaterial(mat);
+          }
+
```
Comment (gonzodepedro) [inline src/systems/collada_world_exporter/ColladaWorldExporter.cc:219]:
```suggestion
scale =  _geom->Data().MeshShape()->Scale();
```

File: test/integration/collada_world_exporter.cc (hunk 29 above)
Comment (chapulina) [reply test/integration/collada_world_exporter.cc:72]: Maybe the working directory could be changed with [common::chdir`](https://github.com/ignitionrobotics/ign-common/blob/a8e59e10b2ad80e7e52523b4869402acbfae5650/include/ignition/common/Filesystem.hh#L159-L162)

File: src/systems/collada_world_exporter/ColladaWorldExporter.cc (hunk 30 above)
Comment (nkoenig) [reply src/systems/collada_world_exporter/ColladaWorldExporter.cc:222]: Added in e44a66efca495ce50beba9ad1b4d173d690ba9be

**Review by azeey — APPROVED** [review]
LGTM!
